File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/listener/impl/EntitySpawnListener.java
package io.github.arcaneplugins.levelledmobs.bukkit.listener.impl;

import io.github.arcaneplugins.levelledmobs.bukkit.LevelledMobs;
import io.github.arcaneplugins.levelledmobs.bukkit.data.InternalEntityDataUtil;
import io.github.arcaneplugins.levelledmobs.bukkit.listener.ListenerWrapper;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context;
import io.github.arcaneplugins.levelledmobs.bukkit.api.data.keys.EntityKeyStore;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.CustomDrop;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.CustomDropHandler;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.cdevent.CustomDropsEventType;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.type.CommandCustomDrop;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.type.ItemCustomDrop;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.type.StandardCustomDropType;
import java.util.Collection;
import javax.annotation.Nonnull;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.entity.EntitySpawnEvent;

public final class EntitySpawnListener extends ListenerWrapper {

    /*
    Constructors
     */

    public EntitySpawnListener() {
        super(true);
    }

    /*
    Methods
     */

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    public void handle(final EntitySpawnEvent event) {

        /*
        LevelledMobs only concerns LivingEntities
         */
        if(!(event.getEntity() instanceof final LivingEntity entity))
            return;

        /*
        Check if the entity has any non-persistent metadata to migrate
         */
        // wasSummoned
        final String wasSummonedKeyStr = EntityKeyStore.WAS_SUMMONED.toString();
        InternalEntityDataUtil.setWasSummoned(
            entity,
            entity.hasMetadata(wasSummonedKeyStr) &&
                entity.getMetadata(wasSummonedKeyStr).stream().anyMatch(val -> val.asInt() == 1),
            true
        );
        entity.removeMetadata(wasSummonedKeyStr, LevelledMobs.getInstance());

        /*
        Fire the associated trigger.
         */
        LogicHandler.runFunctionsWithTriggers(
            new Context().withEntity(entity), "on-entity-spawn"
        );

        /*
        Custom Drops
         */
        handleCustomDrops(event);
    }

    private static void handleCustomDrops(
        final @Nonnull EntitySpawnEvent event
    ) {
        // This is a safe cast since LM will only call this after it has verified this is a LivngEnt
        final LivingEntity entity = (LivingEntity) event.getEntity();

        final Context context = new Context()
            .withEntity(entity)
            .withEvent(event);

        final Collection<CustomDrop> cds =
            CustomDropHandler.getDefinedCustomDropsForEntity(entity);

        for(final @Nonnull CustomDrop cd : cds) {
            if(cd.getType().equals(StandardCustomDropType.ITEM.name())) {
                final ItemCustomDrop icd = (ItemCustomDrop) cd;
                icd.attemptToApplyEquipment(entity);
            } else if(cd.getType().equalsIgnoreCase(StandardCustomDropType.COMMAND.name())) {
                final CommandCustomDrop ccd = (CommandCustomDrop) cd;
                if(ccd.getCommandRunEvents().contains(CustomDropsEventType.ON_SPAWN.name())) {
                    ccd.execute(CustomDropsEventType.ON_SPAWN, context);
                }
            }
        }
    }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/listener/impl/EntitySpawnListener.kt
package io.github.arcaneplugins.levelledmobs.bukkit.listener.impl

import io.github.arcaneplugins.levelledmobs.bukkit.LevelledMobs
import io.github.arcaneplugins.levelledmobs.bukkit.api.data.keys.EntityKeyStore
import io.github.arcaneplugins.levelledmobs.bukkit.data.InternalEntityDataUtil.setSpawnReason
import io.github.arcaneplugins.levelledmobs.bukkit.data.InternalEntityDataUtil.setWasSummoned
import io.github.arcaneplugins.levelledmobs.bukkit.debug.DebugCategory
import io.github.arcaneplugins.levelledmobs.bukkit.debug.DebugHandler
import io.github.arcaneplugins.levelledmobs.bukkit.listener.ListenerWrapper
import io.github.arcaneplugins.levelledmobs.bukkit.logic.LogicHandler.runFunctionsWithTriggers
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.CustomDrop
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.CustomDropHandler.getDefinedCustomDropsForEntity
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.cdevent.CustomDropsEventType
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.type.CommandCustomDrop
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.type.ItemCustomDrop
import io.github.arcaneplugins.levelledmobs.bukkit.logic.customdrops.type.StandardCustomDropType
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.debug
import org.bukkit.entity.LivingEntity
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.entity.CreatureSpawnEvent
import org.bukkit.event.entity.EntitySpawnEvent
import org.bukkit.metadata.MetadataValue

class EntitySpawnListener : ListenerWrapper(true) {
    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun handle(event: EntitySpawnEvent)
    {
        debug(DebugCategory.LISTENERS) {"EntitySpawnListener is handling an event."}

        /*
        LevelledMobs only concerns LivingEntities
         */

        if (event.entity !is LivingEntity) return
        val entity = event.entity as LivingEntity

        /*
        Check if the entity has any non-persistent metadata to migrate
         */
        // wasSummoned
        val wasSummonedKeyStr = EntityKeyStore.WAS_SUMMONED.toString()
        setWasSummoned(entity,entity.hasMetadata(wasSummonedKeyStr) &&
                    entity.getMetadata(wasSummonedKeyStr).stream()
                        .anyMatch { `val`: MetadataValue -> `val`.asInt() == 1 },
            true
        )
        entity.removeMetadata(wasSummonedKeyStr, LevelledMobs.lmInstance)

        /*
        Add other data
         */

        if (event is CreatureSpawnEvent) {
            // Set spawn reason of entity
            setSpawnReason(
                event.entity,
                event.spawnReason,
                true
            )
        }

        /*
        Fire the associated trigger.
         */

        runFunctionsWithTriggers(
            Context(entity),
            mutableListOf("on-entity-spawn")
        )

        /*
        Custom Drops
         */

        handleCustomDrops(event)
    }

    private fun handleCustomDrops(
        event: EntitySpawnEvent
    ){
        // This is a safe cast since LM will only call this after it has verified this is a LivngEnt
        val entity = event.entity as LivingEntity

        val context = Context(entity).withEvent(event)

        val cds: Collection<CustomDrop> = getDefinedCustomDropsForEntity(
            entity,
            context
        )

        for (cd in cds) {
            if (cd.type == StandardCustomDropType.ITEM.name) {
                val icd = cd as ItemCustomDrop
                icd.attemptToApplyEquipment(entity)
            } else if (cd.type.equals(StandardCustomDropType.COMMAND.name, ignoreCase = true)) {
                val ccd = cd as CommandCustomDrop
                if (ccd.commandRunEvents.contains(CustomDropsEventType.ON_SPAWN.name)) {
                    ccd.execute(CustomDropsEventType.ON_SPAWN, context)
                }
            }
        }
    }
}